File: src/main/java/net/hectus/neobb/turn/default_game/structure/TOakDoorTurtling.java
package net.hectus.neobb.turn.default_game.structure;

import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.structure.PlacedStructure;
import net.hectus.neobb.structure.Structure;
import net.hectus.neobb.structure.StructureManager;
import net.hectus.neobb.turn.default_game.CounterFilter;
import net.hectus.neobb.turn.default_game.attributes.clazz.HotClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.NeutralClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.RedstoneClazz;
import net.hectus.neobb.turn.default_game.attributes.function.CounterFunction;
import net.hectus.neobb.turn.default_game.attributes.function.DefenseFunction;
import net.hectus.neobb.util.Modifiers;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import java.util.List;

public class TOakDoorTurtling extends StructureTurn implements DefenseFunction, NeutralClazz {
    public TOakDoorTurtling(NeoPlayer player) { super(player); }
    public TOakDoorTurtling(PlacedStructure data, NeoPlayer player) { super(data, player); }

    @Override
    public int cost() {
        return 4;
    }

    @Override
    public Structure referenceStructure() {
        return StructureManager.structure("oak_door_turtling");
    }

    @Override
    public List<ItemStack> items() {
        // Not sure if it would be OAK_DOOR or some lower part thing with the automatic item detection based on the saved structure.
        return List.of(new ItemStack(Material.OAK_DOOR, 4));
    }

    @Override
    public void applyDefense() {
        player.addModifier(Modifiers.P_DEFAULT_DEFENDED);
        player.game.turnScheduler.runTaskTimer("oak_door_turtling", () -> {
            if (player.game.history().getLast() instanceof CounterFunction &&
                    CounterFilter.clazz(HotClazz.class).doCounter(player.game.history().getLast()) ||
                    CounterFilter.clazz(RedstoneClazz.class).doCounter(player.game.history().getLast())) {
                player.removeModifier(Modifiers.P_DEFAULT_DEFENDED);
            }
        }, () -> !player.hasModifier(Modifiers.P_DEFAULT_DEFENDED), 1);
    }
}

File: src/main/kotlin/net/hectus/neobb/turn/default_game/structure/TOakDoorTurtling.kt
package net.hectus.neobb.turn.default_game.structure

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.game.util.ScheduleID
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.structure.PlacedStructure
import net.hectus.neobb.structure.Structure
import net.hectus.neobb.structure.StructureManager
import net.hectus.neobb.turn.default_game.CounterFilter
import net.hectus.neobb.turn.default_game.attribute.clazz.HotClazz
import net.hectus.neobb.turn.default_game.attribute.clazz.NeutralClazz
import net.hectus.neobb.turn.default_game.attribute.clazz.RedstoneClazz
import net.hectus.neobb.turn.default_game.attribute.function.CounterFunction
import net.hectus.neobb.turn.default_game.attribute.function.DefenseFunction
import net.hectus.neobb.util.Modifiers
import org.bukkit.Material
import org.bukkit.inventory.ItemStack

class TOakDoorTurtling(data: PlacedStructure?, cord: Cord?, player: NeoPlayer?) : StructureTurn(data, cord, player), DefenseFunction, NeutralClazz {
    override val cost: Int = 4
    override val referenceStructure: Structure = StructureManager["oak_door_turtling"]!!

    override fun items(): List<ItemStack> {
        // Not sure if it would be OAK_DOOR or some lower part thing with the automatic item detection based on the saved structure.
        return listOf(ItemStack(Material.OAK_DOOR, 4))
    }

    override fun applyDefense() {
        player!!.addModifier(Modifiers.Player.Default.DEFENDED)
        player.game.turnScheduler.runTaskTimer(ScheduleID.OAK_DOOR_TURTLING, 1, {
            !player.hasModifier(Modifiers.Player.Default.DEFENDED)
        }) {
            if (player.game.history.last() is CounterFunction &&
                CounterFilter.clazz(HotClazz::class).doCounter(player.game.history.last()) ||
                CounterFilter.clazz(RedstoneClazz::class).doCounter(player.game.history.last())
            ) {
                player.removeModifier(Modifiers.Player.Default.DEFENDED)
            }
        }
    }
}